File: EMS/EMS/EMSRootParser.m
//
// Created by Chris Searle on 26/08/14.
// Copyright (c) 2014 Chris Searle. All rights reserved.
//

#import "CJCollection.h"
#import "CJLink.h"

#import "EMSRootParser.h"

@implementation EMSRootParser

- (NSDictionary *)processData:(NSData *)data forHref:(NSURL *)href error:(NSError **)error {
    NSError *parseError = nil;

    CJCollection *collection = [CJCollection collectionForNSData:data error:&parseError];

    if (!collection) {
        DDLogError(@"Failed to retrieve root %@ - %@ - %@", href, parseError, [parseError userInfo]);

        *error = parseError;

        return [NSDictionary dictionary];
    }

    NSMutableDictionary *temp = [[NSMutableDictionary alloc] init];

    [collection.links enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        CJLink *link = (CJLink *) obj;

        temp[link.rel] = link.href;
    }];

    return [NSDictionary dictionaryWithDictionary:temp];
}

- (void)parseData:(NSData *)data forHref:(NSURL *)href {
    NSError *error = nil;

    NSDictionary *collection = [self processData:data forHref:href error:&error];

    [self.delegate finishedRoot:collection forHref:href error:error];
}

@end

File: EMS/Network/Parsers/EMSRootParser.swift
import Foundation
import CocoaLumberjack

@objc protocol EMSRootParserDelegate {
    optional func finishedRoot(links: [String:NSURL], forHref href:NSURL, error:NSError?)
}

@objc class EMSRootParser : NSObject {
    var delegate : EMSRootParserDelegate?
    
    func processData(data: NSData, forHref href:NSURL) throws -> [String:NSURL] {
        
        do {
            let collection = try CJCollection(forNSData:data)
            
            var links : [String:NSURL] = [:]
            
            for link in collection.links {
                links[link.rel] = link.href
            }
            
            return links
        } catch let error as NSError {
            DDLogError("Failed to retrieve root \(href) - \(error)")
            
            throw error
        }
    }
    
    func parseData(data: NSData, forHref url: NSURL) {
        do {
            let collection = try self.processData(data, forHref: url)
            
            self.delegate?.finishedRoot?(collection, forHref: url, error: nil)
        } catch let error as NSError {
            self.delegate?.finishedRoot?([:], forHref: url, error: error)
        }
    }
}
